docs: add detailed instructions for Fedora

keszybz opened the pull request.

keszybz committed:
docs: add detailed instructions for Fedora

lnykryn commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+
+```bash
+KVER=`uname -r`
+dnf download kernel-core-$KVER.`uname -m`
````
In rawhide 
[root@fedora mkosi-initrd]# uname -r
5.15.0-0.rc6.47.fc36.x86_64
So you don't need to append uname -m

lnykryn commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+```bash
+KVER=`uname -r`
+dnf download kernel-core-$KVER.`uname -m`
+sudo PYTHONPATH=$PWD/../mkosi python -m mkosi --default fedora.mkosi -f -o initrd-$KVER.cpio.zst --environment=KERNEL_VERSION=$KVER
````
Again on new clean rawhide I got:
mkosi: error: argument verb: invalid choice: '\nshadow-utils' (choose from 'build', 'clean', 'help', 'summary', 'genkey', 'bump', 'serve', 'build', 'shell', 'boot', 'qemu', 'ssh')

keszybz replied:
I'd guess that you're running some older version of mkosi code, i.e. that the PYTHONPATH is not set correctly. https://github.com/systemd/mkosi/pull/847 should help.

keszybz replied:
Actually no, it's that keszybz/main is behind origin/main. I'll adjust the instructions.

keszybz replied:
I moved ownership to the systemd org.

lnykryn commented on docs/fedora.md:
```diff
@@ -0,0 +1,228 @@
+
+# Building an initramfs image
+
+Two tools are required: `mkosi`, currently a git version is required,
```
surprisingly zstd is not installed by default maybe add it to requirements?

keszybz replied:
I added Recommends to the Fedora package and instructions to this document.

lnykryn commented on docs/fedora.md:
```diff
@@ -0,0 +1,228 @@
+
+Hint: on repeated runs, it may be useful to add --with-network=never if no new packages need to be downloaded.
+
+# Installing the initramfs image
```
For non-efi system 
cp mkosi.output/initrd-$KVER.cpio.zst /boot/
grubby --copy-default --add-kernel=/boot/vmlinuz-$KVER --initrd=/boot/initrd-$KVER.cpio.zst --title="experimental-$KVER" --make-default

keszybz replied:
That's very convenient…

poettering commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+and an output directory to avoid clutter:
+```bash
+mkdir -p mkosi.cache mkosi.output
+```
````
I think it would be nicer to just add

```
…
OutputDirectory=mkosi.output
CacheDirectory=mkosi.cache
…
```

to mkosi.defaults. That way mkosi will create these dirs automatically when they are missing. i.e. it will not just pick them up, but create them. it's a bit like a work-around for the fact that git doesn't store empty dirs...

Benefit: it will just work, and one less step to document

keszybz replied:
Done.

poettering commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+KVER=`uname -r`
+dnf download kernel-core-$KVER.`uname -m`
+sudo PYTHONPATH=$PWD/../mkosi python -m mkosi --default fedora.mkosi -f -o initrd-$KVER.cpio.zst --environment=KERNEL_VERSION=$KVER
+```
````
can we use IMAGE_ID=… + IMAGE_VERSION= for this instead of hardcoding the output name?

The effect would be quite similar, but less manual

keszybz replied:
IMAGE_ID and IMAGE_VERSION would need to be set… I think it's easier to operate with one KVER variable. But maybe I don't understand your idea.

keszybz replied:
I think I get it. I'll try to rework the whole scheme to use ImageId=initrd.

poettering replied:
i'd replace $KVER and $KERNEL_VERSION by $IMAGE_VERSION/--image-version= everywhere. i.e. the cmdline should become:

```console
$ IMAGE_VERSION=5.14.9-300.fc35.x86_64
$ sudo PYTHONPATH=$PWD/../mkosi python -m mkosi --default=fedora.mkosi -f --image-version=$IMAGE_VERSION
$ sudo cp mkosi.output/initrd-$IMAGE_VERSION.cpio.zst /boot/efi/$(cat /etc/machine-id)/$IMAGE_VERSION/initrd
```

poettering replied:
in fact the --default=fedora.mkosi is unnecessary too (given mkosi.default is a symlink to it anyway — why the indirection btw?) Without that it becomes even shorter:

```console
$ IMAGE_VERSION=5.14.9-300.fc35.x86_64
$ sudo PYTHONPATH=$PWD/../mkosi python -m mkosi -f --image-version=$IMAGE_VERSION
$ sudo cp mkosi.output/initrd-$IMAGE_VERSION.cpio.zst /boot/efi/$(cat /etc/machine-id)/$IMAGE_VERSION/initrd
```

poettering replied:
(whatever you specify as ImageId=/ImageVersion=/--image-id=/--image-version= gets passed to the payload build scripts as $IMAGE_ID/$IMAGE_VERSION btw, i.e. the idea is to then use the latter as kernel version to operate with.)

poettering commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+  --format=directory --clean-package-metadata=no
+```
+
+One that's done, we can build a sysext:
````
One → Once

poettering commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+  --base-image=mkosi.output/initrd-$KVER.d
+  --format=gpt_squashfs --environment=SYSEXT=initrd-$KVER-ssh \
+  --package='!*,openssh-server'
+```
````
here too, we probably should use IMAGE_ID/IMAGE_VERSION stuff, i.e. `--image-id=` + `--image-version=`

lnykryn commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+and the `mkosi-initrd` repository with configuration for `mkosi`.
+
+```bash
+git clone https://github.com/keszybz/mkosi
````
I think there should be git clone https://github.com/systemd/mkosi your clone does not have RemovePackages for example

lnykryn commented on docs/fedora.md:
````diff
@@ -0,0 +1,228 @@
+One that's done, we can build a sysext:
+```bash
+sudo PYTHONPATH=$PWD/../mkosi python -m mkosi --default fedora.mkosi -f -o initrd-$KVER-ssh.sysext \
+  --base-image=mkosi.output/initrd-$KVER.d
````
missing \

keszybz committed:
docs: use ImageId= and --image-version=

As suggested in
https://github.com/systemd/mkosi-initrd/pull/1#discussion_r732734308.

keszybz commented:
Updated with all the suggested changes.
There's further stuff to do:
- export `$MKOSI_CACHE` and automatically download the kernel image to that directory
- https://github.com/systemd/mkosi/pull/847 and then remove setting of $PYTHONPATH
- add instructions for signing
- add instructions for sysext installation
